refactor: 10x faster RPC splitting

MarcoPolo opened the pull request:
Builds on #582. 10x faster than current master. 0 allocs.

The basic logic is the same as the old version, except we return an `iter.Seq[RPC]` and yield `RPC` types instead of a slice of `*RPC`. This lets us avoid allocations for heap pointers.

Please review @algorandskiy, and let me know if this improves your use case.

MarcoPolo committed:
add bench

MarcoPolo committed:
Add other benchmark

MarcoPolo committed:
Add other benchmark

MarcoPolo committed:
refactor: 10x faster split RPC logic

MarcoPolo commented on gossipsub.go:
```diff
@@ -1412,135 +1412,182 @@ func (gs *GossipSubRouter) doSendRPC(rpc *RPC, p peer.ID, q *rpcQueue, urgent bo
+// split splits the given RPC If a sub RPC is too large and can't be split
+// further (e.g. Message data is bigger than the RPC limit), then it will be
+// returned as an oversized RPC. The caller should filter out oversized RPCs.
+func (rpc *RPC) split(limit int) iter.Seq[RPC] {
```
This method should likely be in the `pubsub.go` file instead. I kept it here to make reviewing easier, but would like to move it after reviewer's approval.

MarcoPolo committed:
check lint

MarcoPolo commented:
benchstat:

```
goos: darwin
goarch: arm64
pkg: github.com/libp2p/go-libp2p-pubsub
cpu: Apple M1 Max
                                             │   old.txt    │               new.txt               │
                                             │    sec/op    │   sec/op     vs base                │
SplitRPCLargeMessages/Many_large_messages-10   51.751µ ± 0%   4.327µ ± 1%  -91.64% (p=0.000 n=10)
SplitRPCLargeMessages/2_large_messages-10      320.70n ± 2%   34.63n ± 0%  -89.20% (p=0.000 n=10)
geomean                                         4.074µ        387.0n       -90.50%

                                             │   old.txt    │                 new.txt                  │
                                             │     B/op     │    B/op      vs base                     │
SplitRPCLargeMessages/Many_large_messages-10   18.65Ki ± 0%   0.00Ki ± 0%  -100.00% (p=0.000 n=10)
SplitRPCLargeMessages/2_large_messages-10        392.0 ± 0%      0.0 ± 0%  -100.00% (p=0.000 n=10)
geomean                                        2.672Ki                     ?                       ¹ ²
¹ summaries must be >0 to compute geomean
² ratios must be >0 to compute geomean

                                             │  old.txt   │                 new.txt                 │
                                             │ allocs/op  │ allocs/op   vs base                     │
SplitRPCLargeMessages/Many_large_messages-10   217.0 ± 0%     0.0 ± 0%  -100.00% (p=0.000 n=10)
SplitRPCLargeMessages/2_large_messages-10      8.000 ± 0%   0.000 ± 0%  -100.00% (p=0.000 n=10)
geomean                                        41.67                    ?                       ¹ ²
¹ summaries must be >0 to compute geomean
² ratios must be >0 to compute geomean
```

MarcoPolo committed:
lint

algorandskiy commented on gossipsub.go:
```diff
@@ -1412,135 +1412,182 @@ func (gs *GossipSubRouter) doSendRPC(rpc *RPC, p peer.ID, q *rpcQueue, urgent bo
 	}
+}
+
+// pbFieldNumberLT15Size is the number of bytes required to encode a protobuf
```
maybe it makes sense to add a reference to pb.RPC.Size() to let a reader to find the original `1 (pbFieldNumberLT15Size) + l + sov` formula

algorandskiy commented on gossipsub.go:
```diff
@@ -1412,135 +1412,182 @@ func (gs *GossipSubRouter) doSendRPC(rpc *RPC, p peer.ID, q *rpcQueue, urgent bo
-				lastRPC.Subscriptions = append(lastRPC.Subscriptions, sub)
-				out = append(out, lastRPC)
+		for _, sub := range rpc.Subscriptions {
+			if nextRPC.Subscriptions = append(nextRPC.Subscriptions, sub); nextRPC.Size() > limit {
```
`nextRPC.Size()` - this is the same issue with calling Size in a loop over and over again as for Publish but probably does not matter much as (from my understanding) new subscriptions as well as control are relatively rare compare to Publish thing

MarcoPolo replied:
Yes exactly. Notice that we flush the messages to publish above. So here we aren't calling .Size on Publish messages again.

Splitting due to control messages should be extremely rare (happy to be proven wrong), so I opted to do the easier thing and keep the old code.

Thinking about this just now, we could do a fast path check before these loops that check if the original RPC without the publish messages is small enough to fit and can be yielded.

MarcoPolo replied:
Added this, and it adds a nice small improvement. Many large benchmark is around 3.3 us, and 2 large is around 28 ns.

algorandskiy reviewed (approved):
Looks great and impressive +90% perf gain comparing to mine feeb +66% :)

MarcoPolo committed:
Add a fastpath check to see if the RPC without messages is small enough

This assumes it is common for control messages to not be the reason we
have to split.

MarcoPolo committed:
Do not export SizeOfEmbeddedMsg

MarcoPolo committed:
Update comment on pbFieldNumberLT15Size

MarcoPolo committed:
Move RPC.split to pubsub.go file
